Panic on Batch Join-On

I am trying to join across measurements to calculate rates in batch mode.  The measurements have only have one dimension in common so I am trying to join them by using Join-On.  I am using 0.13.1-1.  

When the batch runs, it is encountering the following panic:

```javascript
panic: interface conversion: models.PointInterface is *models.Batch, not models.Batch

goroutine 460 [running]:
panic(0xca7d00, 0xc820c54980)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/influxdata/kapacitor.(*joinset).JoinIntoBatch(0xc82040b7c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /root/go/src/github.com/influxdata/kapacitor/join.go:523 +0xc82
github.com/influxdata/kapacitor.(*group).emitJoinedSet(0xc8204181e0, 0xc82040b7c0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/kapacitor/join.go:383 +0x2a2
github.com/influxdata/kapacitor.(*group).emit(0xc8204181e0, 0x12e9701, 0x0, 0x0)
        /root/go/src/github.com/influxdata/kapacitor/join.go:331 +0x3ce
github.com/influxdata/kapacitor.(*group).collect(0xc8204181e0, 0x0, 0x2b422272b038, 0xc82054f080, 0x0, 0x0)
        /root/go/src/github.com/influxdata/kapacitor/join.go:318 +0xc18
github.com/influxdata/kapacitor.(*group).run(0xc8204181e0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/kapacitor/join.go:256 +0x117
github.com/influxdata/kapacitor.(*JoinNode).getGroup.func1(0xc8204181e0, 0xc8201dfc80, 0xc820319800)
        /root/go/src/github.com/influxdata/kapacitor/join.go:215 +0x30
created by github.com/influxdata/kapacitor.(*JoinNode).getGroup
        /root/go/src/github.com/influxdata/kapacitor/join.go:223 +0x2db
```

The case I am trying is very similar to the Calculate Rates Across Series example from the documentation.

My tick script is as follows:

```javacript
var building = batch
    |query('''
        SELECT sum("value")
        FROM "metricsCenter"."default"."building_power"
     ''')
      .period(24h)
      .every(2m)
      .groupBy(time(5m),'building')
      .fill(0)

var floor = batch
    |query('''
        SELECT sum("value")
        FROM "metricsCenter"."default"."floor_power"
     ''')
      .period(24h)
      .every(2m)
      .groupBy(time(5m),'building','floor')
      .fill(0)

building
    |join(floor)
      .as('building','floor')
      .on('building')
    |log()
      .level('DEBUG')
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,7 +9,8 @@
 ### Bugfixes
 
 - [#656](https://github.com/influxdata/kapacitor/pull/656): Fix issues where an expression could not be passed as a function parameter in TICKscript.
-- [#627](https://github.com/influxdata/kapacitor/issue/327): Fix where InfluxQL functions that returned a batch could drop tags.
+- [#627](https://github.com/influxdata/kapacitor/issue/627): Fix where InfluxQL functions that returned a batch could drop tags.
+- [#674](https://github.com/influxdata/kapacitor/issue/674): Fix panic with Join On and batches.
 
 ## v1.0.0-beta2 [2016-06-17]
 

--- a/integrations/batcher_test.go
+++ b/integrations/batcher_test.go
@@ -100,7 +100,7 @@ batch
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_Derivative", script, 21*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_Derivative", script, 21*time.Second, er, false)
 }
 
 func TestBatch_DerivativeUnit(t *testing.T) {
@@ -147,7 +147,7 @@ batch
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_Derivative", script, 21*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_Derivative", script, 21*time.Second, er, false)
 }
 
 func TestBatch_DerivativeN(t *testing.T) {
@@ -193,7 +193,7 @@ batch
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_DerivativeNN", script, 21*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_DerivativeNN", script, 21*time.Second, er, false)
 }
 
 func TestBatch_DerivativeNN(t *testing.T) {
@@ -236,7 +236,7 @@ batch
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_DerivativeNN", script, 21*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_DerivativeNN", script, 21*time.Second, er, false)
 }
 
 func TestBatch_Elapsed(t *testing.T) {
@@ -269,7 +269,7 @@ batch
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_Elapsed", script, 21*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_Elapsed", script, 21*time.Second, er, false)
 }
 
 func TestBatch_SimpleMR(t *testing.T) {
@@ -324,7 +324,7 @@ batch
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_SimpleMR", script, 30*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_SimpleMR", script, 30*time.Second, er, false)
 }
 func TestBatch_CountEmptyBatch(t *testing.T) {
 	var script = `
@@ -376,7 +376,7 @@ batch
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_CountEmptyBatch", script, 30*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_CountEmptyBatch", script, 30*time.Second, er, false)
 }
 
 func TestBatch_SumEmptyBatch(t *testing.T) {
@@ -430,7 +430,7 @@ batch
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_CountEmptyBatch", script, 30*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_CountEmptyBatch", script, 30*time.Second, er, false)
 }
 
 func TestBatch_GroupBy_TimeOffset(t *testing.T) {
@@ -485,7 +485,7 @@ batch
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_SimpleMR", script, 30*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_SimpleMR", script, 30*time.Second, er, false)
 }
 
 func TestBatch_Default(t *testing.T) {
@@ -522,7 +522,7 @@ batch
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_Default", script, 30*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_Default", script, 30*time.Second, er, false)
 }
 
 func TestBatch_DoubleGroupBy(t *testing.T) {
@@ -556,7 +556,7 @@ batch
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_SimpleMR", script, 30*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_SimpleMR", script, 30*time.Second, er, false)
 }
 
 func TestBatch_AlertAll(t *testing.T) {
@@ -579,7 +579,7 @@ batch
 	// Expect no result since the condition is not met.
 	er := kapacitor.Result{Series: imodels.Rows{}}
 
-	testBatcherWithOutput(t, "TestBatch_SimpleMR", script, 30*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_SimpleMR", script, 30*time.Second, er, false)
 
 	script = `
 batch
@@ -635,7 +635,7 @@ batch
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_SimpleMR", script, 30*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_SimpleMR", script, 30*time.Second, er, false)
 }
 func TestBatch_AlertLevelField(t *testing.T) {
 
@@ -698,7 +698,7 @@ batch
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_SimpleMR", script, 30*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_SimpleMR", script, 30*time.Second, er, false)
 }
 
 func TestBatch_AlertLevelTag(t *testing.T) {
@@ -752,7 +752,7 @@ batch
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_SimpleMR", script, 30*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_SimpleMR", script, 30*time.Second, er, false)
 }
 
 func TestBatch_AlertDuration(t *testing.T) {
@@ -810,7 +810,7 @@ batch
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_SimpleMR", script, 30*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_SimpleMR", script, 30*time.Second, er, false)
 }
 
 func TestBatch_AlertStateChangesOnly(t *testing.T) {
@@ -990,7 +990,7 @@ cpu0
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_Join", script, 30*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_Join", script, 30*time.Second, er, false)
 }
 
 func TestBatch_JoinTolerance(t *testing.T) {
@@ -1041,7 +1041,123 @@ cpu0
 		},
 	}
 
-	testBatcherWithOutput(t, "TestBatch_JoinTolerance", script, 30*time.Second, er)
+	testBatcherWithOutput(t, "TestBatch_JoinTolerance", script, 30*time.Second, er, false)
+}
+
+func TestBatch_JoinOn(t *testing.T) {
+
+	var script = `
+var errorsByServiceGlobal = batch
+	|query('''
+		SELECT sum("value")
+		FROM "telegraf"."default".errors
+''')
+		.period(10s)
+		.every(10s)
+		.groupBy(time(5s),'service')
+
+var errorsByServiceDC = batch
+	|query('''
+		SELECT first("value") as value
+		FROM "telegraf"."default".errors
+''')
+		.period(10s)
+		.every(10s)
+		.groupBy(time(5s),'dc', 'service')
+
+errorsByServiceGlobal
+	|join(errorsByServiceDC)
+		.as('service', 'dc')
+		.on('service')
+		.streamName('dc_error_percent')
+	|eval(lambda: "dc.value" / "service.sum")
+		.keep()
+		.as('value')
+	|httpOut('TestBatch_JoinOn')
+`
+
+	er := kapacitor.Result{
+		Series: imodels.Rows{
+			{
+				Name:    "dc_error_percent",
+				Tags:    map[string]string{"dc": "slc", "service": "cart"},
+				Columns: []string{"time", "dc.value", "service.sum", "value"},
+				Values: [][]interface{}{
+					{
+						time.Date(1971, 1, 1, 0, 0, 0, 0, time.UTC),
+						8.0,
+						11.0,
+						8.0 / 11.0,
+					},
+					{
+						time.Date(1971, 1, 1, 0, 0, 5, 0, time.UTC),
+						3.0,
+						10.0,
+						3.0 / 10.0,
+					},
+				},
+			},
+			{
+				Name:    "dc_error_percent",
+				Tags:    map[string]string{"dc": "nyc", "service": "cart"},
+				Columns: []string{"time", "dc.value", "service.sum", "value"},
+				Values: [][]interface{}{
+					{
+						time.Date(1971, 1, 1, 0, 0, 0, 0, time.UTC),
+						3.0,
+						11.0,
+						3.0 / 11.0,
+					},
+					{
+						time.Date(1971, 1, 1, 0, 0, 5, 0, time.UTC),
+						7.0,
+						10.0,
+						7.0 / 10.0,
+					},
+				},
+			},
+			{
+				Name:    "dc_error_percent",
+				Tags:    map[string]string{"dc": "slc", "service": "login"},
+				Columns: []string{"time", "dc.value", "service.sum", "value"},
+				Values: [][]interface{}{
+					{
+						time.Date(1971, 1, 1, 0, 0, 0, 0, time.UTC),
+						4.0,
+						13.0,
+						4.0 / 13.0,
+					},
+					{
+						time.Date(1971, 1, 1, 0, 0, 5, 0, time.UTC),
+						2.0,
+						8.0,
+						2.0 / 8.0,
+					},
+				},
+			},
+			{
+				Name:    "dc_error_percent",
+				Tags:    map[string]string{"dc": "nyc", "service": "login"},
+				Columns: []string{"time", "dc.value", "service.sum", "value"},
+				Values: [][]interface{}{
+					{
+						time.Date(1971, 1, 1, 0, 0, 0, 0, time.UTC),
+						9.0,
+						13.0,
+						9.0 / 13.0,
+					},
+					{
+						time.Date(1971, 1, 1, 0, 0, 5, 0, time.UTC),
+						6.0,
+						8.0,
+						6.0 / 8.0,
+					},
+				},
+			},
+		},
+	}
+
+	testBatcherWithOutput(t, "TestBatch_JoinOn", script, 30*time.Second, er, true)
 }
 
 // Helper test function for batcher
@@ -1102,7 +1218,7 @@ func testBatcherWithOutput(
 	script string,
 	duration time.Duration,
 	er kapacitor.Result,
-	ignoreOrder ...bool,
+	ignoreOrder bool,
 ) {
 	clock, et, replayErr, tm := testBatcher(t, name, script)
 	defer tm.Close()
@@ -1125,7 +1241,13 @@ func testBatcherWithOutput(
 
 	// Assert we got the expected result
 	result := kapacitor.ResultFromJSON(resp.Body)
-	if eq, msg := compareResults(er, result); !eq {
-		t.Error(msg)
+	if ignoreOrder {
+		if eq, msg := compareResultsIgnoreSeriesOrder(er, result); !eq {
+			t.Error(msg)
+		}
+	} else {
+		if eq, msg := compareResults(er, result); !eq {
+			t.Error(msg)
+		}
 	}
 }

--- a/integrations/data/TestBatch_JoinOn.0.brpl
+++ b/integrations/data/TestBatch_JoinOn.0.brpl
@@ -1,0 +1,4 @@
+{"name":"errors","tags":{"dc":"slc","service":"cart"},"points":[{"fields":{"value":8},"time":"2015-10-30T00:00:00Z"},{"fields":{"value":3},"time":"2015-10-30T00:00:05Z"}]}
+{"name":"errors","tags":{"dc":"nyc","service":"cart"},"points":[{"fields":{"value":3},"time":"2015-10-30T00:00:00Z"},{"fields":{"value":7},"time":"2015-10-30T00:00:05Z"}]}
+{"name":"errors","tags":{"dc":"slc","service":"login"},"points":[{"fields":{"value":4},"time":"2015-10-30T00:00:00Z"},{"fields":{"value":2},"time":"2015-10-30T00:00:05Z"}]}
+{"name":"errors","tags":{"dc":"nyc","service":"login"},"points":[{"fields":{"value":9},"time":"2015-10-30T00:00:00Z"},{"fields":{"value":6},"time":"2015-10-30T00:00:05Z"}]}

--- a/integrations/data/TestBatch_JoinOn.1.brpl
+++ b/integrations/data/TestBatch_JoinOn.1.brpl
@@ -1,0 +1,2 @@
+{"name":"errors","tags":{"service":"cart"},"points":[{"fields":{"sum":11},"time":"2015-10-30T00:00:00Z"},{"fields":{"sum":10},"time":"2015-10-30T00:00:05Z"}]}
+{"name":"errors","tags":{"service":"login"},"points":[{"fields":{"sum":13},"time":"2015-10-30T00:00:00Z"},{"fields":{"sum":8},"time":"2015-10-30T00:00:05Z"}]}

--- a/join.go
+++ b/join.go
@@ -200,7 +200,7 @@ func (j *JoinNode) sendMatchPoint(specific, matched srcPoint, groupErrs chan<- e
 	// Send current point
 	group.points <- specific
 	// Send new matched point
-	matched.p = np
+	matched.p = np.Interface()
 	group.points <- matched
 }
 
@@ -520,7 +520,11 @@ func (js *joinset) JoinIntoBatch() (models.Batch, bool) {
 				empty[i] = true
 				continue
 			}
-			b := batch.(models.Batch)
+			b, ok := batch.(models.Batch)
+			if !ok {
+				js.logger.Printf("E! invalid join data got %T expected models.Batch", batch)
+				return models.Batch{}, false
+			}
 			if indexes[i] == len(b.Points) {
 				emptyCount++
 				empty[i] = true

--- a/models/batch.go
+++ b/models/batch.go
@@ -79,6 +79,10 @@ func (b Batch) Setter() PointSetter {
 	return &b
 }
 
+func (b *Batch) Interface() PointInterface {
+	return *b
+}
+
 func (b *Batch) SetNewDimTag(key string, value string) {
 	b.Tags[key] = value
 	for _, p := range b.Points {
@@ -86,7 +90,9 @@ func (b *Batch) SetNewDimTag(key string, value string) {
 	}
 }
 
-func (b *Batch) UpdateGroup() {}
+func (b *Batch) UpdateGroup() {
+	b.Group = TagsToGroupID(b.PointDimensions(), b.Tags)
+}
 
 func BatchToRow(b Batch) (row *models.Row) {
 	row = &models.Row{

--- a/models/point.go
+++ b/models/point.go
@@ -35,6 +35,7 @@ type PointSetter interface {
 	PointInterface
 	SetNewDimTag(key string, value string)
 	UpdateGroup()
+	Interface() PointInterface
 }
 
 // Represents a single data point
@@ -91,6 +92,10 @@ func (p Point) Copy() PointInterface {
 
 func (p Point) Setter() PointSetter {
 	return &p
+}
+
+func (p *Point) Interface() PointInterface {
+	return *p
 }
 
 func (p *Point) SetNewDimTag(key string, value string) {

--- a/pipeline/node.go
+++ b/pipeline/node.go
@@ -229,7 +229,7 @@ const intervalMarker = "INTERVAL"
 //        |alert()
 //            .id('node \'stream0\' in task \'{{ .TaskName }}\'')
 //            .message('{{ .ID }} is {{ if eq .Level "OK" }}alive{{ else }}dead{{ end }}: {{ index .Fields "emitted" | printf "%0.3f" }} points/10s.')
-//            .crit(lamdba: "emitted" <= 100.0)
+//            .crit(lambda: "emitted" <= 100.0)
 //    //Do normal processing of data
 //    data...
 //

